src/mlpack/methods/lsh/lsh_search_impl.hpp

```diff
@@ -181,12 +187,225 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
         referenceIndex, distance);
 }
 
+// Compare class for <double, size_t> pair, used in GetAdditionalProbingBins
+class CompareGreater
+{
+  public:
+    bool operator()(
+        std::pair<double, size_t> p1, 
+        std::pair<double, size_t> p2){
+      //only compare the double values
+      return p1.first > p2.first;
+    }
+};
+
+//Returns the score of a perturbation vector generated by perturbation set A
+//The score of a pertubation set (vector) is the sum of scores of the
+//participating actions
+inline double perturbationScore(
+              const std::vector<size_t> &A,
+              const arma::vec &scores)
+{
+  double score = 0.0;
+  for (size_t i = 0; i < A.size(); ++i)
+    score+=scores[i];
+  return score;
+}
+
+// Replace max element with max element+1 in perturbation set A
+inline void perturbationShift(std::vector<size_t> &A, const size_t numProj)
+{
+  size_t max_pos = 0;
+  size_t max = A[0];
+  for (size_t i = 1; i < A.size(); ++i)
+  {
+    if (A[i] > max)
+    {
+      max = A[i];
+      max_pos = i;
+    }
+  }
+  if (A[max_pos] < 2*numProj - 1)
+    A[max_pos]++;
+  else
+    return; //don't add indices that will make A invalid
+}
+
+// Add 1+max element to perturbation set A
+inline void perturbationExpand(std::vector<size_t> &A, const size_t numProj)
+{
+  size_t max = A[0];
+  for (size_t i = 1; i < A.size(); ++i)
+    if (A[i] > max)
+      max = A[i];
+  if (max+1 >= 2*numProj){
+    return; //don't add indices that will make A invalid
+  }
+  A.push_back(max+1);
+}
+
+// Return true if perturbation set A is valid. A perturbation set is invalid if
+// it contains two (or more) actions for the same dimension or dimensions that
+// are larger than the queryCode's dimensions.
+inline bool perturbationValid(
+    const std::vector<size_t> &A,
+    const arma::Col<size_t> &positions,
+    const size_t numProj)
+{
+  bool check[numProj] = {0};
+
+  for (size_t i = 0; i < A.size(); ++i)
+  {
+    if ( positions[A[i]] >= numProj )
+      return false; //invalid if contains non-existing dimension
+
+    //check that we only see each dimension once
+    if (check[A[i]] == 0)
+      check[A[i]] = 1;
+    else
+      return false;
+  }
+  return true;
+}
+
+
+// Compute additional probing bins for a query
+template<typename SortPolicy>
+void LSHSearch<SortPolicy>::GetAdditionalProbingBins(
+                            const arma::vec &queryCode,
+                            const arma::vec &queryCodeNotFloored,
+                            const size_t T,
+                            arma::mat &additionalProbingBins) const
+{
+
+  if (T == 0)
+    return;
+
+  // Each column of additionalProbingBins is the code of a bin.
+  additionalProbingBins.zeros(numProj, T);
```

Comment:
.set_size() will avoid the time to set everything to zero, which will be useful here since you are setting the columns in the for loop below this call.